### ISSUE TITLE
state.sls / state.apply does NOT honor commandline 'pillar={...}' values

### ISSUE THREAD

**Issue by tkwilliams** [saltstack/salt#39627]
The docs state that
```
Pillar data can be set on the command line when running state.apply <salt.modules.state.apply_() like so:

salt '*' state.apply pillar='{"foo": "bar"}'
salt '*' state.apply my_sls_file pillar='{"hello": "world"}'

Nested pillar values can also be set via the command line:

salt '*' state.sls my_sls_file pillar='{"foo": {"bar": "baz"}}'
```
It appears this DOES work for SOME invocations (e.g. 'pillar.items', 'state.show_sls') but flatly fails if invoked in the style in the docs.  None of the values passed in the pillar dict are available during the state run, even though an IDENTICAL call substituting 'state.show_sls' for 'state.sls' works perfectly and produces the expected behavior and output.

In particular I have a mako file with the following:
```
<%!
  from salt.exceptions import SaltInvocationError
%>\
<%
  bootstrap_vpc = pillar.get('bootstrap_vpc')
  try:
    product, environment, loc = bootstrap_vpc.split('-')
    role = 'salt-master'
  except:
    msg = ("Invalid value for bootstrap_vpc passed ('%s').  When bootstrapping a new environment, "
           "bootstrap_vpc must be of the form 'product-env-location' where 'product' is the 3 letter "
           "product identifier, 'env' is the 2-3 char environment type, and 'location' is the 4-5 "
           "char provider / region specifier." % bootstrap_vpc)
    raise SaltInvocationError(msg)
    return ''
%>
...more stuff here...
```

This works perfectly when I run e.g. 
```
salt-call pillar.items pillar='{bootstrap_vpc: plt-prd-aue1}'
```
or even
```
salt-call state.show_sls classes.aws.contents pillar='{bootstrap_vpc: plt-prd-aue1}'
```
but if I then attempt to run
```
salt-call state.sls classes.aws.contents pillar='{bootstrap_vpc: plt-prd-aue1}'
```
'pillar.get('bootstrap_vpc')' suddenly becomes 'None' and the except clause is called.

This is putting a BIG dent in a big implementation I'm on tap to deliver, so any thoughts or workarounds would be mucho appreciated

**Comment by tkwilliams** [comment]
I should also note that, if I bypass the try/except and simply hardcode bootstrap_vpc to some valid value, everything in the state run works perfectly, so it's not an issue with the pillar, state, or makostack files - it's strictly a failure of salt to pass the pillar values from the commandline along, when state.sls is invoked...

**Comment by tkwilliams** [comment]
OK, I believe I've tracked the actual issue down - it appears that the ext_pillar() functions in ext_pillar modules are called TWICE, the FIRST time with an empty pillar dict, EVEN IF pillar data was provided on the command line.  The SECOND time, the provided pillar data is passed correctly.  Sadly, my code expects (nee requires) the params to be there as soon as the ext_pillar (makostack in this case) is loaded, leading to the issues describe above.

I've implemented a workaround in the makostack ext_pillar module (since that's my particular use-case / pain point) which I'll be dropping a PR for shortly, but I'd bet more than even money that this will affect ANY ext_pillar module in the tree :-/

Again, it's REALLY whacky that this bug is only exercised by state.sls, but not by state.show_sls ...

@cachedout sorry to tap you on the shoulder :) but this is pretty certainly a bug in the ext_pillar implementation and I don't seem to be able to directly flag it as such.

**Comment by terminalmage** [comment]
Versions report?

**Comment by terminalmage** [comment]
Also, please provide a full case that can be used to replicate this (master config, makostack config, commands to run, etc.). I'm not at all familiar with the makostack ext_pillar, and your partial example isn't enough for me to figure it out in conjunction with the existing docs without spending a lot of extra time I don't have to figure it out.

I am familiar with how pillar works internally though, and I would like to find out why it would be calling this ext_pillar twice and fix the root cause.

**Comment by tkwilliams** [comment]
Sorry, wasn't really thinking.  It's current develop as of a couple days ago
```
root@inf-dev-salt-aue1-01# salt --versions-report
Salt Version:
           Salt: 2016.11.0

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
          gitdb: 0.6.4
      gitpython: 2.0.8
          ioflo: Not Installed
         Jinja2: unknown
        libgit2: 0.20.0
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.4
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: 0.20.3
         Python: 2.7.13 (default, Dec 20 2016, 15:20:00)
   python-gnupg: 0.3.8
         PyYAML: 3.10
          PyZMQ: 14.3.1
           RAET: Not Installed
          smmap: 0.8.1
        timelib: Not Installed
        Tornado: 4.4.1
            ZMQ: 3.2.5

System Versions:
           dist: centos 6.8 Final
        machine: x86_64
        release: 2.6.32-642.11.1.el6.x86_64
         system: Linux
        version: CentOS 6.8 Final
```

**Comment by tkwilliams** [comment]
My actual setup is a bit ... convoluted?  Largely due to the need to support our developers building against multiple products, production states and isolated micro-application trees.

I'll try to work up a minimal config that exercises the bug in question and attach it, hopefully tomorrow as it's quite late here already....

**Comment by terminalmage** [comment]
Yeah, a minimal config to reproduce the issue is what I was after.

It's late here too, so don't worry. We can pick this back up on Friday or whenever is convenient.

**Comment by tkwilliams** [comment]
Awesome, thanks!

Have a good evening.

**Comment by tkwilliams** [comment]
[39627.tar.gz](https://github.com/saltstack/salt/files/805060/39627.tar.gz)

A very bare /etc/salt and /srv/* setup to generate the mentioned error.

Calling e.g.
```
salt-call state.sls foo pillar='{bootstrap_vpc: plt-dev-aue1}'
```
will generate an error, while
```
salt-call state.show_sls foo pillar='{bootstrap_vpc: plt-dev-aue1}'
```
or
```
salt-call pillar.items pillar='{bootstrap_vpc: plt-dev-aue1}'
```
will not.

I've noticed that an error gets logged to /var/log/salt/master in BOTH cases, but only `state.sls` appears to pass the error up to the top-level where it fails the run.  The stock code does not log a successful parsing of makostack yaml files, only on various failure modes, so it's not as easy to see that they're being parsed twice (once without any specified pillar values, and once with) -- I recommend applying the following to make it easier to track
```
--- /usr/lib/python2.7/site-packages/salt/pillar/makostack.py	2017-02-27 16:40:29.652722065 -0500
+++ /usr/lib/python2.7/site-packages/salt/pillar/makostack.py.bak	2017-02-27 16:40:14.558104345 -0500
@@ -470,6 +470,7 @@
                 for sub in namespace.split(':')[::-1]:
                     obj = {sub: obj}
             stack = _merge_dict(stack, obj)
+            log.info('Stack template {} parsed'.format(path))
         except exceptions.TopLevelLookupException as e:
             log.info('Stack template "{0}" not found.'.format(path))
             continue
```

**Comment by terminalmage** [comment]
@tkwilliams I've found the issue. ``state.sls`` is looking for pillar render errors and aborting before trying to run the states if any are found. However, it was checking the [in-memory pillar data](https://docs.saltstack.com/en/latest/topics/pillar/#in-memory-pillar-data-vs-on-demand-pillar-data) instead of the pillar data generated on-demand when the states are run.

As a bit of history, ``state.sls``, ``state.highstate``, and other functions which run states used to operate on the in-memory pillar data, which is only compiled on minion start, and when ``saltutil.refresh_pillar`` is run. But what would happen is people would change their pillar data and run states, and not get the new pillar data. So at some point, we modified the state compiler so that it always compiles fresh pillar data for each state run, leaving the in-memory pillar data unchanged.

However, it looks like we did not update the code that checks for pillar render errors. The code was looking for render errors in the in-memory pillar data, which was last compiled on minion startup and did not have the CLI pillar override in place. So, that error you're seeing in the logs is 100% expected behavior.

Because we were looking in the wrong place for render errors, and because this check happened before we would have tried to gather fresh pillar data for state execution, Salt was aborting prematurely from ``state.sls`` with the error you were seeing. It wasn't even getting to the point where it was even _trying_ to run the SLS file.

I've opened https://github.com/saltstack/salt/pull/39727, which changes the code which checks for pillar render errors so that it checks the freshly-gathered pillar. I also expanded error checking to the other functions which instantiate a ``salt.state.HighState`` class instance, so that we get better error reporting in those functions as well.

I recommend closing https://github.com/saltstack/salt/pull/39630 since https://github.com/saltstack/salt/pull/39727 will fix the underlying issue. However, it would be great if you could open a new pull request containing the documentation/logging fixes from https://github.com/saltstack/salt/pull/39630.

**Comment by terminalmage** [comment]
Also, as a bit of clarification, we are _not_ compiling pillar twice when running states. It looks like you were mistaking the pillar compile that happens on minion startup and thinking it was happening as a result of running ``state.sls``, etc. You should be able to view the timestamps in the master log to confirm this.

**Comment by tkwilliams** [comment]
Extra awesome!  Thanks @terminalmage - I'll fixup the PRs as you suggest, and keep an eye out for #39727 to come through.  As soon as I see it, I'll pull and verify it resolves my problem.

Thanks again!
t.

**Comment by terminalmage** [comment]
@tkwilliams no problem. I submitted my PR against 2016.11 so it will be merged into the develop branch within a couple days of it being merged.

**Comment by terminalmage** [comment]
And thanks for taking the time to set up that test case, it really helped. I bundled that case into a docker container.

1. Launch the container using ``docker run --rm -it -v /path/to/git/clone:/testing terminalmage/issues:39627``.
2. Start the master and minion using ``salt-master -d; salt-minion -d``
3. Run your salt-call commands.

Using this container will allow you to test salt from a git checkout, so you can either [check out the pull request directly from GitHub](https://help.github.com/articles/checking-out-pull-requests-locally/) or wait for it to merge into develop and just test using the head of the develop branch.

If I grok the use case you set up, it should just return ``True`` from that SLS file and not produce any state results when you run the ``state.sls``. It looks like the point of the test case was to make sure that the error was not raised. Please do correct me if I'm wrong, I want to make sure we get this right.

**Comment by terminalmage** [comment]
Hmm, it looks like the Docker Hub is still having issues today, my push only partially completed. If that docker image doesn't show up, give it a while and try again.

**Comment by tkwilliams** [comment]
OK, got the extra-super-latest (with the sys.path fix) but sadly the issue mentioned in this thread is not fixed.  I still get no commandline pillar values during the ext_pillar() call into makostack.  I've not dug far enough to say whether it still shows two runs of the module (one without pillar and one with).  But given that the error is the same, and the same workaround (pasted below) still fixes the problem for me, as before...
```
diff --git a/salt/pillar/makostack.py b/salt/pillar/makostack.py
index e42be2a..6888971 100644
--- a/salt/pillar/makostack.py
+++ b/salt/pillar/makostack.py
@@ -407,6 +407,8 @@ def __virtual__():


 def ext_pillar(minion_id, pillar, *args, **kwargs):
+    if pillar == {}:
+        return {'ext_pillar': 'makostack'}
     import salt.utils
     stack = {}
     stack_config_files = list(args)
```

The purpose of the patch was to bypass the first of the two runs - since it's called with an empty pillar, I set SOMETHING in there and return immediately.  When ext_pillar() is called the second time, pillar will have something in it (either from the commandline, or at least the 'ext_pillar': 'makostack' I set the first time around) and thus knows it's safe to continue.

As mentioned it's an ugly hack, but does allow things to work when the command line pillar data is not being passed that first time through...

**Comment by tkwilliams** [comment]
Ideally, the pillar data from the CL would be passed in every time ext_pillar() is called, removing the need for the above hack....

**Comment by tkwilliams** [comment]
Ahhh, it looks like it was merged with 2016.11 instead of develop?  That would likely be the source of the continued problem, since I'm building my RPMs from develop.  Gonna try cherry-picking your changes onto my working branch and see if it helps...

**Comment by tkwilliams** [comment]
@terminalmage -- I can confirm that cherry-picking your changes into develop (with some minor conflict resolution) fixes the issue.  The makostack ext_pillar now loads nicely with no errors, and no workaround required - with command line pillar available immediately as expected.  :)

Thanks very much!

@cachedout  Any thoughts on when this might be ported to develop?

**Comment by terminalmage** [comment]
@tkwilliams Yes, we use a merge-forward workflow in which bugfixes are ideally submitted to the earliest non-EOL branch where the bug exists. This prevents the need for frequent backporting of individual issues, as we regularly merge forward from older release branches into newer release branches. This way, a bugfix submitted against the 2016.3 branch end up being merged into the 2016.11 branch, and then the 2016.11 branch would be merged into the develop branch.

I would expect the 2016.11 branch to be merged forward into develop early next week.

More on our workflow can be found [here](https://docs.saltstack.com/en/latest/topics/development/contributing.html#which-salt-branch).
